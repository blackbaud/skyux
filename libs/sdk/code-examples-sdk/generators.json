{
  "generators": {
    "code-example-spa": {
      "factory": "./src/generators/code-example-spa/generator",
      "schema": "./src/generators/code-example-spa/schema.json",
      "description": "Generates a container SPA for a code example"
    },
    "build": {
      "factory": "./src/generators/build/generator",
      "schema": "./src/generators/build/schema.json",
      "description": "Generates a set of code examples"
    }
  }
}
