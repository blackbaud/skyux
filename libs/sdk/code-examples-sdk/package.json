{
  "name": "@skyux-sdk/code-examples",
  "version": "0.0.1",
  "type": "commonjs",
  "generators": "./generators.json"
}
