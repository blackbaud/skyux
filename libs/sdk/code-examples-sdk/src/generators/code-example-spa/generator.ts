import { VERSION as ANGULAR_VERSION } from '@angular/cli';
import {
  Tree,
  formatFiles,
  generateFiles,
  updateJson,
  workspaceRoot,
} from '@nx/devkit';
import { dasherize } from '@nx/devkit/src/utils/string-utils';
import { Schema as NgNewSchema } from '@schematics/angular/ng-new/schema';
import { findNodes } from '@schematics/angular/utility/ast-utils';

import { wrapAngularDevkitSchematic } from 'nx/src/adapter/ngcli-adapter';
import { getHoistedPackageVersion } from 'nx/src/plugins/js/lock-file/utils/package-json';
import { readPackageJson } from 'nx/src/project-graph/file-utils';
import * as ts from 'typescript';

import {
  findComponentClass,
  findNgModuleClass,
  readSourceFile,
} from '../../utils/ast-utils';
import {
  findDependenciesFromCode,
  findPeerDependencies,
} from '../../utils/find-dependencies';

import {
  CodeExampleSpaGeneratorConfig,
  CodeExampleSpaGeneratorSchema,
} from './schema';

const codeExamplesBasePath = `apps/code-examples/src/app/code-examples`;
const buildPath = `dist/libs/sdk/code-examples-sdk`;
const commonJsDependencies = {
  autonumeric: ['@skyux/ag-grid', '@skyux/autonumeric'],
  moment: ['@skyux/ag-grid', '@skyux/datetime'],
  'dom-autoscroller': ['@skyux/lookup', '@skyux/lists'],
};
const startingDependencies = [
  '@skyux/assets',
  '@skyux/core',
  '@skyux/i18n',
  '@skyux/icons',
  '@skyux/packages',
  '@skyux/theme',
];

function normalizeOptions(
  options: CodeExampleSpaGeneratorSchema
): CodeExampleSpaGeneratorConfig {
  return {
    ...options,
    project: `${dasherize(options.component)}-${dasherize(
      options.path.replaceAll('/', '-')
    )}`,
    projectPath: `${dasherize(options.component)}/${dasherize(options.path)}`,
  };
}

function findExampleModule(tree: Tree, config: CodeExampleSpaGeneratorConfig) {
  // Get the module file
  const exampleModuleFiles = tree
    .children(`${codeExamplesBasePath}/${config.projectPath}`)
    .filter((child) => child.endsWith('.module.ts'));
  if (exampleModuleFiles.length !== 1) {
    throw new Error(
      `Expected exactly one module file for ${config.projectPath} but found ${
        exampleModuleFiles.length
      }: ${exampleModuleFiles.join(', ')}`
    );
  }
  const exampleModuleFile = exampleModuleFiles[0];

  // Get the module class name
  const exampleModuleSource = readSourceFile(
    tree,
    `${codeExamplesBasePath}/${config.projectPath}/${exampleModuleFile}`
  );
  const exampleModuleClass = findNgModuleClass(exampleModuleSource);
  const exampleModuleClassName =
    exampleModuleClass?.classDeclaration.name?.getText();

  // Get the example component class name
  const exampleModuleExportsArray = exampleModuleClass?.properties[
    'exports'
  ] as ts.ArrayLiteralExpression;
  const exampleModuleExportIdentifiers =
    exampleModuleExportsArray.elements.filter(
      ts.isIdentifier
    ) as ts.Identifier[];
  const exampleComponentClassIdentifier = exampleModuleExportIdentifiers.filter(
    (identifier) => identifier.getText().endsWith('Component')
  )[0];
  const exampleComponentClassName = exampleComponentClassIdentifier.getText();

  // Get the import path for the example component.
  const exampleComponentImport = findNodes(
    exampleModuleSource,
    ts.SyntaxKind.ImportDeclaration
  ).filter((node) =>
    (node as ts.ImportDeclaration).getText().includes(exampleComponentClassName)
  )[0] as ts.ImportDeclaration;
  if (
    !exampleModuleClass ||
    !exampleModuleClassName ||
    !exampleComponentImport
  ) {
    throw new Error(
      `Could not find import for example component in ${config.projectPath}/${exampleModuleFile}`
    );
  }
  let exampleComponentFile = exampleComponentImport.moduleSpecifier.getText();
  // Remove quotes, add .ts extension.
  exampleComponentFile =
    exampleComponentFile.substring(1, exampleComponentFile.length - 1) + '.ts';

  // Get the example component source.
  const exampleComponentSource = readSourceFile(
    tree,
    `${codeExamplesBasePath}/${config.projectPath}/${exampleComponentFile}`
  );

  // Get the example component selector.
  const exampleComponent = findComponentClass(exampleComponentSource);
  if (!exampleComponent) {
    throw new Error(
      `Could not find Component decorator in ${config.projectPath}/${exampleComponentFile}`
    );
  }
  const exampleComponentSelector = (
    exampleComponent.properties['selector'] as ts.StringLiteral
  ).text;

  const hasTests = tree
    .children(`${codeExamplesBasePath}/${config.projectPath}`)
    .some((child) => child.endsWith('.spec.ts'));

  return {
    exampleComponentSelector,
    exampleModuleFile,
    exampleModuleClassName,
    hasTests,
  };
}

async function generateSpa(
  config: CodeExampleSpaGeneratorConfig,
  tree: Tree,
  exampleModuleClassName: string,
  exampleModuleFile: string,
  dependencies: string[],
  exampleComponentSelector: string,
  hasTests: boolean
) {
  const ngNew = wrapAngularDevkitSchematic('@schematics/angular', 'ng-new');
  const cdkAdd = wrapAngularDevkitSchematic('@angular/cdk', 'ng-add');
  const outputPath = `${buildPath}/for-github/${config.projectPath}`;

  if (tree.isFile(`${outputPath}/angular.json`)) {
    throw new Error(
      `The project build ${outputPath} already exists. Please delete it before running this schematic.`
    );
  }

  await ngNew(tree, {
    name: config.project,
    directory: outputPath,
    skipGit: true,
    skipInstall: true,
    skipTests: !hasTests,
    packageManager: 'npm',
    routing: false,
    strict: true,
    style: 'scss',
    version: ANGULAR_VERSION.full,
  } as NgNewSchema);
  await cdkAdd(tree, {
    project: config.project,
  });

  tree.delete(`${outputPath}/README.md`);
  tree
    .children(`${outputPath}/src/app`)
    .forEach((child) => tree.delete(`${outputPath}/src/app/${child}`));
  tree.delete(`${outputPath}/src/assets/.gitkeep`);
  tree.delete(`${outputPath}/src/favicon.ico`);

  // Update angular.json
  updateJson(tree, `${outputPath}/angular.json`, (json) => {
    delete json.projects[config.project].architect.build.configurations;
    delete json.projects[config.project].architect.build.defaultConfiguration;
    delete json.projects[config.project].architect.serve.configurations;
    delete json.projects[config.project].architect.serve.defaultConfiguration;
    if (!json.projects[config.project].architect.serve.options) {
      json.projects[config.project].architect.serve.options = {};
    }
    json.projects[
      config.project
    ].architect.serve.options.browserTarget = `${config.project}:build`;
    json.projects[config.project].architect.build.options.polyfills.push(
      '@skyux/packages/polyfills'
    );
    json.projects[config.project].architect.build.options.assets = [];
    json.projects[config.project].architect.build.options.buildOptimizer =
      false;
    json.projects[config.project].architect.build.options.optimization = false;
    json.projects[
      config.project
    ].architect.build.options.allowedCommonJsDependencies = [
      'fontfaceobserver',
      '@skyux/icons',
    ];
    Object.entries(commonJsDependencies).forEach(
      ([dependency, skyuxModules]) => {
        if (skyuxModules.some((module) => dependencies.includes(module))) {
          json.projects[
            config.project
          ].architect.build.options.allowedCommonJsDependencies.push(
            dependency
          );
        }
      }
    );

    if (hasTests) {
      json.projects[config.project].architect.test.options.polyfills.push(
        '@skyux/packages/polyfills'
      );
      json.projects[config.project].architect.test.options.assets = [];
    } else {
      delete json.projects[config.project].architect.test;
    }
    delete json.projects[config.project].schematics;
    return json;
  });

  // Update package.json
  updateJson(tree, `${outputPath}/package.json`, (json) => {
    const newDependencies = dependencies.filter((dependency) => {
      return (
        !json.dependencies[dependency] && !json.devDependencies[dependency]
      );
    });
    const skyuxVersion = `^${readPackageJson().version}`;
    newDependencies.forEach((dependency) => {
      if (dependency.startsWith('@skyux')) {
        if (tree.isFile(`node_modules/${dependency}/package.json`)) {
          json.dependencies[dependency] = `${getHoistedPackageVersion(
            dependency
          )}`;
        } else {
          json.dependencies[dependency] = skyuxVersion;
        }
      } else {
        json.dependencies[dependency] = `${getHoistedPackageVersion(
          dependency
        )}`;
      }
    });
    const sortedDependencies: Record<string, string> = {};
    Object.keys(json.dependencies)
      .sort((a, b) => {
        if (a.startsWith('@') && !b.startsWith('@')) {
          return -1;
        } else if (!a.startsWith('@') && b.startsWith('@')) {
          return 1;
        } else {
          return a.localeCompare(b, undefined, { sensitivity: 'base' });
        }
      })
      .forEach((key) => {
        sortedDependencies[key] = json.dependencies[key];
      });
    json.dependencies = sortedDependencies;
    if (!hasTests) {
      delete json.scripts.test;
      [
        '@types/jasmine',
        'jasmine-core',
        'karma',
        'karma-chrome-launcher',
        'karma-coverage',
        'karma-jasmine',
        'karma-jasmine-html-reporter',
      ].forEach((dependency) => {
        delete json.devDependencies[dependency];
      });
    }
    return json;
  });

  // Update tsconfig.json
  updateJson(tree, `${outputPath}/tsconfig.json`, (json) => {
    delete json.compileOnSave;
    json.compilerOptions.esModuleInterop = true;
    json.angularCompilerOptions.strictTemplates = true;
    return json;
  });

  if (!hasTests) {
    tree.delete(`${outputPath}/tsconfig.spec.json`);
  }

  // Append to .gitignore
  const gitignorePath = `${outputPath}/.gitignore`;
  const gitignore = tree.read(gitignorePath, 'utf-8');
  tree.write(gitignorePath, `${gitignore}\n\n.angular\n`);

  // Overwrite files with our own
  generateFiles(tree, `${__dirname}/files/project`, outputPath, {
    name: config.project,
    exampleModuleClassName,
    exampleModuleFile: `${exampleModuleFile}`.replace('.ts', ''),
    selector: exampleComponentSelector,
    projectPath: config.projectPath,
    dasherize,
    includeAgGrid: dependencies.includes('ag-grid-angular'),
  });
  return outputPath;
}

function generateHtmlLauncherFile(
  tree: Tree,
  outputPath: string,
  config: CodeExampleSpaGeneratorConfig
) {
  // Generate an HTML launcher file.
  const files: Record<string, string> = {};
  const getFiles = (path: string) => {
    tree.children(path).forEach((child) => {
      const filePath = `${path}/${child}`;
      if (tree.isFile(filePath)) {
        files[filePath.replace(`${outputPath}/`, '')] = `${tree.read(
          filePath,
          'utf-8'
        )}`;
      } else {
        getFiles(filePath);
      }
    });
  };
  getFiles(outputPath);
  generateFiles(tree, `${__dirname}/files/launcher`, `${buildPath}/launchers`, {
    files,
    name: config.project,
    jsonStringify: JSON.stringify,
  });
}

export async function codeExampleSpa(
  tree: Tree,
  options: CodeExampleSpaGeneratorSchema
) {
  const config = normalizeOptions(options);
  const {
    exampleModuleFile,
    exampleModuleClassName,
    exampleComponentSelector,
    hasTests,
  } = findExampleModule(tree, config);
  const exampleCodeDependencies = findDependenciesFromCode(
    tree,
    `${codeExamplesBasePath}/${config.projectPath}`
  );
  const exampleDependencies = findPeerDependencies(
    tree,
    exampleCodeDependencies.concat(startingDependencies)
  );

  // Create the basic SPA, including dependencies.
  const outputPath = await generateSpa(
    config,
    tree,
    exampleModuleClassName,
    exampleModuleFile,
    exampleDependencies,
    exampleComponentSelector,
    hasTests
  );

  // Copy code example component
  generateFiles(
    tree,
    `${workspaceRoot}/${codeExamplesBasePath}/${config.projectPath}`,
    `${outputPath}/src/app`,
    {}
  );

  // Format before generating launcher file.
  await formatFiles(tree);

  // Generate an HTML launcher file.
  generateHtmlLauncherFile(tree, outputPath, config);

  // Format again after generating launcher file.
  await formatFiles(tree);
}

export default codeExampleSpa;
