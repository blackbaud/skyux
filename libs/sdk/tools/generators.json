{
  "generators": {
    "update-dependencies": {
      "factory": "./src/generators/update-dependencies/generator",
      "schema": "./src/generators/update-dependencies/schema.json",
      "description": "Update peer dependencies, such as during a major version upgrade"
    }
  }
}
