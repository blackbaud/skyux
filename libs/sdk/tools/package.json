{
  "name": "@skyux-sdk/tools",
  "generators": "./generators.json",
  "peerDependencies": {
    "@nx/devkit": "^20.0.0"
  }
}
