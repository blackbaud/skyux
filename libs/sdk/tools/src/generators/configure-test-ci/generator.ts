import {
  Tree,
  formatFiles,
  getProjects,
  updateProjectConfiguration,
} from '@nx/devkit';

export async function configureTestCiGenerator(
  tree: Tree,
  options: { skipFormat: boolean },
) {
  const projects = getProjects(tree);
  projects.forEach((project, projectName) => {
    if (project.targets?.['test']) {
      if (
        project.targets['test'].executor ===
        '@angular-devkit/build-angular:karma'
      ) {
        project.targets['test'].configurations = {
          ...project.targets['test'].configurations,
          ci: {
            browsers: 'ChromeHeadlessNoSandbox',
            codeCoverage: true,
            progress: false,
            sourceMap: true,
            watch: false,
          },
        };
      } else if (project.targets['test'].executor === '@nx/jest:jest') {
        project.targets['test'].configurations = {
          ...project.targets['test'].configurations,
          ci: {
            ci: true,
            codeCoverage: true,
            runInBand: true,
          },
        };
      }
    }
    updateProjectConfiguration(tree, projectName, project);
  });
  /* istanbul ignore next */
  if (!options.skipFormat) {
    await formatFiles(tree);
  }
}

export default configureTestCiGenerator;
