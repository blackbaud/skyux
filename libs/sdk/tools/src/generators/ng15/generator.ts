import {join, Path, relative} from '@angular-devkit/core';
import {
  Tree,
  formatFiles,
  getProjects,
  updateProjectConfiguration,
  updateJson, readJson,
} from '@nrwl/devkit';
import * as semver from 'semver';

import { TargetConfiguration } from 'nx/src/config/workspace-json-project-json';
import {NgPackageConfig} from "ng-packagr/ng-package.schema";
import { ok } from 'node:assert';

function setStylesheetImportPathOnLibraries(tree: Tree) {
  getProjects(tree).forEach((projectConfig) => {
    if (
      projectConfig.name &&
      projectConfig.projectType === 'library' &&
      projectConfig.targets?.build?.executor === '@nrwl/angular:package' &&
      projectConfig.targets?.build?.options?.project.includes('ng-package.json') &&
      tree.isFile(projectConfig.targets.build.options.project)
    ) {
      updateJson(tree, projectConfig.targets.build.options.project, (json: NgPackageConfig) => {
        if (!json.inlineStyleLanguage) {
          json.inlineStyleLanguage = 'scss';
        }
        if (json.inlineStyleLanguage === 'scss') {
          const root = <Path>'/';
          const pathToRoot = relative(join(root, projectConfig.root), root);
          ok(json.lib, 'ng-package.json must have a lib property');
          json.lib.styleIncludePaths = json.lib.styleIncludePaths || [];
          if (!json.lib.styleIncludePaths.includes(pathToRoot)) {
            json.lib.styleIncludePaths.push(pathToRoot);
          }
        }
        return json;
      });
    }
  });
}

function addPolyfillsToTarget(target: TargetConfiguration) {
  if (!target.options.polyfills) {
    target.options.polyfills = [];
  } else if (!(target.options?.polyfills instanceof Array)) {
    target.options.polyfills = [target.options.polyfills];
  }
  ['zone.js', 'zone.js/testing'].forEach((polyfill) => {
    if (!target.options.polyfills.includes(polyfill)) {
      target.options.polyfills.push(polyfill);
    }
  });
}

function addPolyfillsToProjectConfigTestTargets(tree: Tree) {
  getProjects(tree).forEach((projectConfig) => {
    if (projectConfig.name) {
      if (
        projectConfig.targets?.test?.executor ===
        '@angular-devkit/build-angular:karma'
      ) {
        addPolyfillsToTarget(projectConfig.targets.test);
      }
      updateProjectConfiguration(tree, projectConfig.name, projectConfig);
    }
  });
}

function removePolyfillsFromTestTs(tree: Tree) {
  getProjects(tree).forEach((projectConfig) => {
    if (projectConfig.name) {
      if (
        projectConfig.targets?.test?.executor ===
        '@angular-devkit/build-angular:karma'
      ) {
        const testTsPath = `${projectConfig.root}/src/test.ts`;
        if (tree.exists(testTsPath)) {
          const testTs = `${tree.read(testTsPath, 'utf-8')}`;
          const originalTestTs = testTs;
          const updatedTestTs = testTs.replace(
            /import 'zone.js\/(dist\/)?zone(-testing)?';\n?/g,
            ''
          );
          if (originalTestTs !== updatedTestTs) {
            tree.write(testTsPath, updatedTestTs);
          }
        }
      }
    }
  });
}

function updatePeerDependencies(tree: Tree) {
  const rootPackageJson = readJson(tree, 'package.json');
  const rootPackageJsonDependencies = {
    ...rootPackageJson.dependencies,
    ...rootPackageJson.devDependencies,
  };
  getProjects(tree).forEach((projectConfig) => {
    if (tree.exists(`${projectConfig.root}/package.json`)) {
      updateJson(tree, `${projectConfig.root}/package.json`, (json) => {
        ['dependencies', 'peerDependencies'].forEach((dependencyType) => {
          if (dependencyType in json) {
            const dependencies = json[dependencyType];
            for (const [name, version] of Object.entries(dependencies)) {
              if (name in rootPackageJsonDependencies) {
                const rootVersion = rootPackageJsonDependencies[name];
                if (`${rootVersion}`.match(/^[0-9]+\.[0-9]+\.[0-9]+$/)) {
                  if (dependencyType === 'peerDependencies' && `${version}`.match(/^\^[0-9]+\.[0-9]+\.[0-9]+$/)) {
                    dependencies[name] = `^${rootVersion}`;
                  } else if (!semver.satisfies(rootVersion, `${version}`)) {
                    throw new Error(
                      `The version of ${name} in ${projectConfig.root} (${version}) is not compatible with the version ${rootVersion} from the root package.json.`
                    );
                  }
                }
              }
            }
          }
        });
        return json;
      });
    }
  });
}

export default async function (tree: Tree) {
  setStylesheetImportPathOnLibraries(tree);
  addPolyfillsToProjectConfigTestTargets(tree);
  removePolyfillsFromTestTs(tree);
  updatePeerDependencies(tree);
  await formatFiles(tree);
}
