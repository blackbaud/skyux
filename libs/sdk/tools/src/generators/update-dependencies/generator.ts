import {
  Tree,
  formatFiles,
  getProjects,
  readJson,
  updateJson,
  updateProjectConfiguration,
  visitNotIgnoredFiles,
} from '@nx/devkit';
import { getSourceNodes } from '@nx/js';

import * as semver from 'semver';
import * as ts from 'typescript';

function hardenRootDependencies(tree: Tree): void {
  updateJson(tree, 'package.json', (json) => {
    ['dependencies', 'devDependencies'].forEach((dependencyType) => {
      if (dependencyType in json) {
        const dependencies = json[dependencyType];
        for (const [name, version] of Object.entries<string>(dependencies)) {
          if (`${version}`.match(/^[~^][0-9]+\.[0-9]+\.[0-9]+$/)) {
            json[dependencyType][name] = version.substring(1);
          }
        }
      }
    });
    return json;
  });
}

function updatePeerDependencies(tree: Tree): void {
  const rootPackageJson = readJson(tree, 'package.json');
  const rootPackageJsonDependencies: Record<string, string> = {
    ...rootPackageJson.dependencies,
    ...rootPackageJson.devDependencies,
  };
  getProjects(tree).forEach((projectConfig) => {
    if (tree.exists(`${projectConfig.root}/package.json`)) {
      updateJson(tree, `${projectConfig.root}/package.json`, (json) => {
        ['dependencies', 'peerDependencies'].forEach((dependencyType) => {
          if (dependencyType in json) {
            const dependencies = json[dependencyType];
            for (const [name, version] of Object.entries(dependencies)) {
              if (name in rootPackageJsonDependencies) {
                const rootVersion = rootPackageJsonDependencies[name];
                if (`${rootVersion}`.match(/^[0-9]+\.[0-9]+\.[0-9]+$/)) {
                  if (
                    dependencyType === 'peerDependencies' &&
                    `${version}`.match(/^\^[0-9]+\.[0-9]+\.[0-9]+$/)
                  ) {
                    json[dependencyType][name] = `^${rootVersion}`;
                  } else if (!semver.satisfies(rootVersion, `${version}`)) {
                    throw new Error(
                      `The version of ${name} in ${projectConfig.root} (${version}) is not compatible with the version ${rootVersion} from the root package.json.`,
                    );
                  }
                }
              }
            }
          }
        });
        return json;
      });
    }
  });
}

function updateNgUpdatePackageGroup(tree: Tree): void {
  const rootPackageJson = readJson(tree, 'package.json');
  const rootPackageJsonDependencies: Record<string, string> = {
    ...rootPackageJson.dependencies,
    ...rootPackageJson.devDependencies,
  };
  if (tree.exists('libs/components/packages/package.json')) {
    updateJson(tree, 'libs/components/packages/package.json', (json) => {
      const packageGroup = Object.entries<string>(
        json['ng-update']['packageGroup'],
      );
      packageGroup.forEach(([name, version]) => {
        if (version !== '0.0.0-PLACEHOLDER') {
          const rootVersion = rootPackageJsonDependencies[name];
          const prefix = version.match(`^[~^]`) ? version.charAt(0) : '';
          if (rootVersion) {
            json['ng-update']['packageGroup'][name] = `${prefix}${rootVersion}`;
          }
        }
      });
      return json;
    });
  }
}

/**
 * Testing projects do not have a separate package.json to track dependencies.
 */
function updateTestingImplicitDependencies(tree: Tree): void {
  const projects = getProjects(tree);
  const testingProjects = new Map(
    Array.from(projects.entries()).filter(([name]) =>
      name.endsWith('-testing'),
    ),
  );
  const packagesToProjects = new Map<string, string>();
  projects.forEach((projectConfig) => {
    const packageJsonFile = `${projectConfig.root}/package.json`;
    if (projectConfig.name && tree.exists(packageJsonFile)) {
      const packageJson = readJson(tree, packageJsonFile);
      if (packageJson.name) {
        if (testingProjects.has(`${projectConfig.name}-testing`)) {
          packagesToProjects.set(
            `${packageJson.name}/testing`,
            `${projectConfig.name}-testing`,
          );
        }
        packagesToProjects.set(packageJson.name, projectConfig.name);
      }
    }
  });
  const packages = Array.from(packagesToProjects.keys());
  testingProjects.forEach((projectConfig) => {
    const importedProjects: string[] = [];
    visitNotIgnoredFiles(tree, `${projectConfig.root}/src`, (file) => {
      if (file.endsWith('.ts')) {
        const source = ts.createSourceFile(
          file,
          tree.read(file, 'utf-8') || '',
          ts.ScriptTarget.Latest,
          true,
        );
        const sourceNodes = getSourceNodes(source);
        const importDeclarations = sourceNodes.filter((node) =>
          ts.isImportDeclaration(node),
        ) as ts.ImportDeclaration[];
        importDeclarations.forEach((importDeclaration) => {
          const importPath = importDeclaration.moduleSpecifier.getText();
          const importProject = packages.find(
            (packageName) =>
              importPath === `'${packageName}'` ||
              importPath.startsWith(`'${packageName}/`),
          );
          const projectName = packagesToProjects.get(importProject ?? '');
          if (importProject && projectName) {
            importedProjects.push(projectName);
          }
        });
      }
    });
    if (projectConfig.name) {
      projectConfig.targets ??= {};
      projectConfig.targets['build'] = {
        command: `echo ' 🏗️  build ${projectConfig.name}'`,
        dependsOn: [
          {
            projects: [projectConfig.name.replace('-testing', '')],
            target: 'build',
          },
        ],
      };
      if (importedProjects.length > 0) {
        projectConfig.implicitDependencies = [
          ...new Set(importedProjects),
        ].sort((a, b) => a.localeCompare(b));
      }
      updateProjectConfiguration(tree, projectConfig.name, projectConfig);
    }
  });
}

export default async function (tree: Tree) {
  hardenRootDependencies(tree);
  updatePeerDependencies(tree);
  updateNgUpdatePackageGroup(tree);
  updateTestingImplicitDependencies(tree);
  await formatFiles(tree);
}
