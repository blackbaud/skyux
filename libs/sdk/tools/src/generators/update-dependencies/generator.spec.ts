import { version as ANGULAR_VERSION } from '@angular/core/package.json';
import { libraryGenerator } from '@nx/angular/generators';
import {
  Tree,
  readJson,
  readProjectConfiguration,
  stripIndents,
  updateJson,
  writeJson,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';

import generator from './generator';

describe('update dependencies generator', () => {
  let appTree: Tree;

  beforeEach(() => {
    appTree = createTreeWithEmptyWorkspace();
    writeJson(appTree, 'package.json', {
      dependencies: {
        '@angular/common': ANGULAR_VERSION,
        '@angular/core': ANGULAR_VERSION,
      },
    });
    appTree.write('.gitignore', 'node_modules');
    writeJson(appTree, '.prettierrc', {
      singleQuote: true,
      importOrder: ['^@(.*)$', '^\\w(.*)$', '^(../)(.*)$', '^(./)(.*)$'],
      importOrderSeparation: true,
      importOrderSortSpecifiers: true,
      importOrderParserPlugins: ['typescript', 'decorators-legacy'],
    });
    appTree.write('.prettierignore', `.prettierrc\njest.preset.js\n`);
  });

  it('should run successfully', async () => {
    await libraryGenerator(appTree, {
      name: 'test',
      buildable: true,
      publishable: true,
      importPath: '@proj/test',
      directory: 'test',
      skipFormat: true,
      skipPackageJson: true,
    });
    updateJson(appTree, 'package.json', (json: any) => {
      json.dependencies['example-package'] = '^1.0.0';
      return json;
    });
    await generator(appTree);
    const config = readProjectConfiguration(appTree, 'test');
    expect(config).toBeDefined();
    expect(
      readJson(appTree, `package.json`).dependencies['example-package'],
    ).toEqual('1.0.0');
  });

  it('should update peer dependencies', async () => {
    await libraryGenerator(appTree, {
      name: 'test',
      buildable: true,
      publishable: true,
      importPath: '@proj/test',
      directory: 'test',
      skipFormat: true,
      skipPackageJson: true,
    });
    const originalConfig = readProjectConfiguration(appTree, 'test');
    expect(originalConfig).toBeDefined();
    updateJson(appTree, 'package.json', (json: any) => {
      json.dependencies = {
        ...json.dependencies,
        '@proj/one': '1.0.0',
      };
      json.devDependencies = {
        ...json.devDependencies,
        '@proj/two': '2.0.0',
      };
      return json;
    });
    updateJson(appTree, `${originalConfig.root}/package.json`, (json: any) => {
      json.dependencies = {
        ...json.dependencies,
        '@proj/one': '^1.0.0',
      };
      json.peerDependencies = {
        ...json.peerDependencies,
        '@proj/two': '^1.0.0',
      };
      return json;
    });
    await generator(appTree);
    const projectPackageJson = readJson(
      appTree,
      `${originalConfig.root}/package.json`,
    );
    expect(projectPackageJson).toBeDefined();
    expect(projectPackageJson.peerDependencies).toEqual({
      '@angular/common': `^${ANGULAR_VERSION}`,
      '@angular/core': `^${ANGULAR_VERSION}`,
      '@proj/two': '^2.0.0',
    });
  });

  it('should update ng-update package group', async () => {
    appTree.write(
      'package.json',
      JSON.stringify({
        dependencies: {
          '@proj/one': '1.1.0',
        },
        devDependencies: {
          '@proj/two': '2.2.0',
        },
      }),
    );
    appTree.write(
      'libs/components/packages/package.json',
      JSON.stringify({
        'ng-update': {
          packageGroup: {
            '@proj/one': '1.0.0',
            '@proj/two': '^2.0.0',
          },
        },
      }),
    );
    await generator(appTree);
    const projectPackageJson = readJson(
      appTree,
      `libs/components/packages/package.json`,
    );
    expect(projectPackageJson).toEqual({
      'ng-update': {
        packageGroup: {
          '@proj/one': '1.1.0',
          '@proj/two': '^2.2.0',
        },
      },
    });
  });

  it('should throw error on unmet dependencies', async () => {
    await libraryGenerator(appTree, {
      name: 'test',
      buildable: true,
      publishable: true,
      importPath: '@proj/test',
      directory: 'test',
      skipFormat: true,
      skipPackageJson: true,
    });
    const originalConfig = readProjectConfiguration(appTree, 'test');
    expect(originalConfig).toBeDefined();
    updateJson(appTree, 'package.json', (json: any) => {
      json.dependencies = {
        ...json.dependencies,
        '@proj/one': '2.0.0',
      };
      return json;
    });
    updateJson(appTree, `${originalConfig.root}/package.json`, (json: any) => {
      json.dependencies = {
        ...json.dependencies,
        '@proj/one': '^1.0.0',
      };
      return json;
    });
    await expect(() => generator(appTree)).rejects.toThrow(
      new Error(
        `The version of @proj/one in test (^1.0.0) is not compatible with the version 2.0.0 from the root package.json.`,
      ),
    );
  });

  it('should capture testing dependencies', async () => {
    await libraryGenerator(appTree, {
      name: 'test',
      buildable: true,
      publishable: true,
      importPath: '@proj/test',
      directory: 'lib/test',
      skipFormat: true,
      skipPackageJson: true,
    });
    await libraryGenerator(appTree, {
      name: 'other',
      buildable: true,
      publishable: true,
      importPath: '@proj/other',
      directory: 'lib/other',
      skipFormat: true,
      skipPackageJson: true,
    });
    const testProject = readProjectConfiguration(appTree, 'test');
    writeJson(appTree, `${testProject.root}/testing/project.json`, {
      name: 'test-testing',
      projectType: 'library',
      sourceRoot: `${testProject.root}/testing/src`,
      targets: {
        test: {
          commands: ['echo test'],
        },
      },
    });
    appTree.write(`${testProject.root}/testing/src/empty.ts`, '');
    appTree.write(
      `${testProject.root}/testing/src/test.ts`,
      stripIndents`
      import { thirdParty } from '@third-party/core';
      import { test } from '@proj/test';
      import { other } from '@proj/other';

      test();
      other();
      `,
    );
    await generator(appTree);
    expect(readProjectConfiguration(appTree, 'test')).toMatchSnapshot();
    expect(readProjectConfiguration(appTree, 'test-testing')).toMatchSnapshot();
  });
});
