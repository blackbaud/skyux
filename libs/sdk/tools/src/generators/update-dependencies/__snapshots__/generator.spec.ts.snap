// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update dependencies generator should capture testing dependencies 1`] = `
Object {
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "name": "test",
  "prefix": "lib",
  "projectType": "library",
  "root": "lib/test",
  "sourceRoot": "lib/test/src",
  "tags": Array [],
  "targets": Object {
    "build": Object {
      "configurations": Object {
        "development": Object {
          "tsConfig": "lib/test/tsconfig.lib.json",
        },
        "production": Object {
          "tsConfig": "lib/test/tsconfig.lib.prod.json",
        },
      },
      "defaultConfiguration": "production",
      "dependsOn": Array [
        "^build",
        Object {
          "projects": Array [
            "aaa",
            "bbb",
          ],
          "target": "build",
        },
      ],
      "executor": "@nx/angular:package",
      "inputs": Array [
        "buildInputs",
        "^buildInputs",
        "{workspaceRoot}/lib/test/testing/src/**/*",
        "!{workspaceRoot}/lib/test/testing/src/**/*.spec.ts",
        "!{workspaceRoot}/lib/test/testing/src/**/fixtures/**/*",
      ],
      "options": Object {
        "project": "lib/test/ng-package.json",
      },
      "outputs": Array [
        "{workspaceRoot}/dist/{projectRoot}",
      ],
    },
    "lint": Object {
      "executor": "@nx/eslint:lint",
    },
    "test": Object {
      "executor": "@nx/jest:jest",
      "options": Object {
        "jestConfig": "lib/test/jest.config.ts",
      },
      "outputs": Array [
        "{workspaceRoot}/coverage/{projectRoot}",
      ],
    },
  },
}
`;

exports[`update dependencies generator should capture testing dependencies 2`] = `
Object {
  "name": "test-testing",
  "projectType": "library",
  "root": "lib/test/testing",
  "sourceRoot": "lib/test/testing/src",
  "targets": Object {
    "test": Object {
      "commands": Array [
        "echo test",
      ],
    },
  },
}
`;
